src/Val.cc

```diff
@@ -4262,9 +4262,77 @@ ValPtr cast_value_to_type(Val* v, Type* t)
 		return static_cast<Broker::detail::DataVal*>(dv.get())->castTo(t);
 		}
 
+	// Allow casting between sets and vectors if the yield types are the same.
+	if ( v->GetType()->IsSet() && IsVector(t->Tag()) )
+		{
+		auto set_type = v->GetType<SetType>();
+		auto indices = set_type->GetIndices();
+
+		if ( indices->GetTypes().size() > 1 || ! indices->IsPure() )
+			return nullptr;
+
+		auto ret_type = make_intrusive<VectorType>(indices->GetPureType());
+		auto ret = make_intrusive<VectorVal>(ret_type);
+
+		auto lv = v->AsTableVal()->ToPureListVal();
+		auto n = lv->Length();
+
+		for ( int i = 0; i < n; ++i )
+			ret->Assign(i, lv->Idx(i));
+
+		return ret;
+		}
+	else if ( IsVector(v->GetType()->Tag()) && t->IsSet() )
+		{
+		auto tlp = make_intrusive<TypeList>(v->GetType()->Yield());
+		tlp->Append(v->GetType()->Yield());
+
+		auto ret_type = make_intrusive<SetType>(tlp, nullptr);
```

Comment:
Can't `ret_type` be replaced with the type of the cast expression `t`? Same for the VectorType above.

Otherwise every converted value has a new instance of a type which seems unnecessary.